### PR TITLE
Leader Election Metrics 2a: Server side wiring 

### PR DESCRIPTION
**Pull request by Jolyon-S**
**Goals (and why)**:
* Add a new endpoint to the Timelock feedback service to allow for reporting of leader election metrics.
* Add wiring to aggregate received metrics

**Implementation Description (bullets)**:
* Added a new endpoint
* Added a class to aggregate received metrics

**Testing (What was existing testing like?  What have you done to improve it?)**:
* WIP

**Concerns (what feedback would you like?)**:
* I'm only adding a new endpoint - no requests will be sent yet. I assume this is fine?

**Where should we start reviewing?**:
* `timelock-api.yml` is probably a good start.

**Priority (whenever / two weeks / yesterday)**:
After https://github.com/palantir/atlasdb/pull/5069 merges.

### CODE REVIEW

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.atlasdb.util.SlidingWindowWeightedMeanGauge;
+import com.palantir.conjure.java.lib.SafeLong;
+import com.palantir.timelock.feedback.LeaderElectionStatistics;
+
+public final class LeaderElectionMetricAggregator {
+    private final SlidingWindowWeightedMeanGauge weightedGaugeP99;
+    private final SlidingWindowWeightedMeanGauge weightedGaugeP95;
+    private final SlidingWindowWeightedMeanGauge weightedGaugeMean;
+    private volatile long lastElectionTime = 0L;
```
Comment (Jolyon-S) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java:29]: I'm not a huge fan, but not sure what the best way to use a gauge here is.

File: lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java [hunk 2]
```diff
@@ -197,13 +230,17 @@ private void assertExpectedDuration(Instant instant, Instant instant2) {
...
+        @Value.Parameter
         long responseMillis();
 
-        @Parameter
+        @Value.Parameter
         UUID responseLeader();
+
+        static SingleCall of(long requestMillis, long responseMillis, UUID responseLeader) {
+            return ImmutableSingleCall.of(requestMillis, responseMillis, responseLeader);
+        }
```
Comment (Jolyon-S) [inline lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java:244]: Driveby

Comment (gmaretic) [reply lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java:244]: Have fun rebasing :P

File: timelock-api/src/main/conjure/timelock-feedback.yml [hunk 3]
```diff
@@ -17,6 +17,13 @@ types:
           p99: double
           oneMin: double
           errorRate: optional<double>
+      LeaderElectionStatistics:
+        fields:
+          p99: double
+          p95: double
+          mean: double
+          count: safelong
+          perceivedTime: optional<safelong>
```
Comment (gmaretic) [inline timelock-api/src/main/conjure/timelock-feedback.yml:26]: durationEstimate?

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                LeaderElectionMetricAggregator.class, "leaderElectionImpactP99", weightedGaugeP99);
+        metricsManager.registerMetric(LeaderElectionMetricAggregator.class, "leaderElectionDuration", lastElectionTime);
+    }
+
+    void report(LeaderElectionStatistics statistics) {
+        long count = statistics.getCount().longValue();
+        weightedGaugeMean.update(statistics.getMean(), count);
+        weightedGaugeP95.update(statistics.getP95(), count);
+        weightedGaugeP99.update(statistics.getP99(), count);
+        lastElectionTime.setValue(
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java:50]: Might be better to just not set when empty

File: lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java [hunk 5]
```diff
@@ -115,6 +120,24 @@ public ConjureStartTransactionsResponse startTransactions(ConjureStartTransactio
...
+    }
+
+    @VisibleForTesting
+    LeaderElectionStatistics statisticsWithRegistry(TaggedMetricRegistry metricRegistry) {
+        Snapshot metricsSnapshot = metrics.observedDuration().getSnapshot();
+        LeaderElectionStatistics electionStatistics = LeaderElectionStatistics.builder()
+                .p99(metricsSnapshot.get99thPercentile())
+                .p95(metricsSnapshot.get95thPercentile())
+                .mean(metricsSnapshot.getMean())
+                .count(SafeLong.of(metrics.observedDuration().getCount()))
```
Comment (gmaretic) [inline lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java:134]: There is of course a small race condition here, but it's fine

File: lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java [hunk 6]
```diff
@@ -115,6 +120,24 @@ public ConjureStartTransactionsResponse startTransactions(ConjureStartTransactio
                 .logId());
     }
 
+    public LeaderElectionStatistics statistics() {
+        return statisticsWithRegistry(new DefaultTaggedMetricRegistry());
+    }
+
+    @VisibleForTesting
+    LeaderElectionStatistics statisticsWithRegistry(TaggedMetricRegistry metricRegistry) {
```
Comment (gmaretic) [inline lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java:128]: This is a bit janky, but probably fine for tests. Rename to `getStatisticsAndSetRegistryTo` or something such?

File: lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java (hunk 2 above)
Comment (Jolyon-S) [reply lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java:244]: I will, don't worry

**Review by gmaretic — COMMENTED** [review]
LGTM, a few nits and needs rebasing on the base branch, but other than that 👍

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java (hunk 4 above)
Comment (Jolyon-S) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java:50]: Reasoning here being that, when no election is going on, we might want to set to 0 (else the metric will sit on the previous observed value between elections, which might give the impression that we're having a bad time)

File: lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java (hunk 6 above)
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java:128]: Yea, I hate this but it does make it much easier to test.

File: lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java (hunk 5 above)
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java]: Did some digging and I believe that `metricSnapshot.size()` is equivalent, just without the race condition, so replacing with that.

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java (hunk 4 above)
Comment (gmaretic) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionMetricAggregator.java]: Yeah, still not buying it as that's exactly what the metric is supposed to do: report the last known leader election, and 0 will skew the picture

**Review by gmaretic — APPROVED** [review]

**Comment by Jolyon-S** [comment]
:+1: seems legit

**Pull request merged**